zkevm-circuits/src/state_circuit/constraint_builder.rs

```diff
@@ -226,10 +231,80 @@ impl<F: Field> ConstraintBuilder<F> {
         // TODO: Missing constraints
     }
 
+    fn build_tx_log_constraints(&mut self, q: &Queries<F>) {
+        self.require_equal(
+            "is_write is always true for TxLog",
+            q.is_write.clone(),
+            1.expr(),
+        );
+        self.require_in_set(
+            "field_tag in TxLogFieldTag range",
+            q.field_tag(),
+            set::<F, TxLogFieldTag>(),
+        );
+        // reset log_id when tx_id increases
+        let no_tag_change = 1.expr() - q.tag_change();
```

Comment:
is q.tag_change() always boolean? if not, is the no_tag_change correct?

Reply:
oh, it should be not always boolean even after sort by tag if some tag not appear, looks like need to add another column to check the diff.

Reply:
add is_tag_unchanged  in https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/552/commits/3017da9d864edd3d8e5b47fee97f34c94bb1f5e6 thanks !!